### ISSUE TITLE
Skip tax calculation on $0 orders

### ISSUE THREAD

**Issue by devinsays** [taxjar/taxjar-woocommerce-plugin#173]
Our store generates a large amount of $0 subscription renewals every month. These are gift subscriptions that have an upfront (prepaid) signup fee, and then renew every month for a specific number of months.

The [calculate_renewal_order_totals](https://github.com/taxjar/taxjar-woocommerce-plugin/blob/master/includes/class-wc-taxjar-integration.php#L1028) method is run for each of these renewals, which calls `calculate_order_tax` to make an API call to TaxJar, and then update the order totals.

Because we process such a large amount of $0 orders (75,000+ on a single day ever month), it would be great if we could avoid the performance hit of an API call and tax calculations.

There's a couple of options for implementation.

### Option A

We could specifically exclude these $0 orders directly in the `calculate_renewal_order_totals` and `calculate_order_tax` methods, which would benefit everyone processing $0 orders who uses this TaxJar plugin.

```
// Skip tax calculation order total is 0.
if ( 0 == $order->get_total() ) {
    return;
}
```

I think of a case where a shop would need to calculate tax on a $0 order, but maybe it exists?

### Option B

We could add filters to the `calculate_renewal_order_totals` and `calculate_order_tax` methods, which would allow developers to eject from tax calculation whenever an order meets a specific criteria. This would allow us to return early if the order is $0, and also provide flexibility for other use cases developers may have.

```
// Calculate TaxJar taxes for renewal order unless overridden by filter.
if ( ! apply_filters( 'taxjar_subscription_renewal_order_should_calculate_tax', true, $order, $subscription ) ) {
    return $order;
}
```

```
// Calculate TaxJar taxes for order unless overridden by filter.
if ( ! apply_filters( 'taxjar_order_should_calculate_tax', true, $order ) ) {
    return;
}
```

I can't think of a downside for this option.

### Option C

Combine Option A + Option B, to automatically exclude $0 orders from tax calculation and provide the additional flexibility in case any developers need it down the line.

--

I plan to create a PR for Option C, but can limit it to Option A or Option B if you have concerns about any of the implementations.

Let me know if you have any questions or feedback!

**Comment by dallendalton** [comment]
@devinsays - Thank you for submitting your suggestion! We've looked over it and believe option C would be a good improvement to our plugin. 

We have added to our roadmap and look forward to your PR. (Even if you don't submit a PR we will get make this improvement). 

Also, you may want to add an additional filter and the order total logic in the [calculate_backend_totals](https://github.com/taxjar/taxjar-woocommerce-plugin/blob/17b3f63b1a98ae21ff021ae2c45c874dda48f41c/includes/class-wc-taxjar-integration.php#L966) method, as this method is used to calculate an order's tax when it is created directly in the admin.

Thanks!

**Comment by devinsays** [comment]
@dallendalton Thanks for the feedback.

I just created a draft PR:
https://github.com/taxjar/taxjar-woocommerce-plugin/pull/174

I'll add a filter for `calculate_backend_totals` as suggested.

I realized the 0 check is not required in `calculate_order_tax` since this method is only called from two places: `calculate_api_order_tax` and `calculate_renewal_order_totals`. So, I just added the check earlier and prevent the call to `calculate_order_tax` altogether. However, I did leave a `taxjar_order_should_calculate_tax` filter in.

Working on unit tests now, which will take a bit longer.
